File: Components/DevExpressVCL/Demos/ExpressPivotGrid/CBuilder/UnboundExternalDemo/UnboundExternalMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "UnboundExternalMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxLookAndFeels"
#pragma link "DemoBasicMain"
#pragma link "cxControls"
#pragma link "cxCustomPivotGrid"
#pragma link "cxDBPivotGrid"
#pragma link "cxPivotGrid"
#pragma resource "*.dfm"
TfrmUnboundExternal *frmUnboundExternal;
//---------------------------------------------------------------------------
__fastcall TfrmUnboundExternal::TfrmUnboundExternal(TComponent* Owner)
        : TfrmDemoBasicMain(Owner)
{
}

void __fastcall TfrmUnboundExternal::FormCreate(TObject* Sender)
{
  TfrmDemoBasicMain::FormCreate(Sender); 
  TcxExternalDataSource* ExternalData = new TcxExternalDataSource("ExternalData.dat");
  PivotGrid()->BeginUpdate();
  try {
    for (int I = 0; I < ExternalData->FieldCount; I++) {
      TcxPivotGridField* AField = PivotGrid()->CreateField();
      AField->Caption = ExternalData->FieldNames[I];
      AField->DataBinding->ValueType = ExternalData->FieldTypes[I];
      AField->Visible = True;
    }
    SetFieldPos("PaymentType", faColumn);
    SetFieldPos("Payment Amount", faData);
    SetFieldPos("Quantity", faData);
    SetFieldPos("Company Name", faRow);
    SetFieldPos("Car Name", faRow);
    PivotGrid()->DataController->CustomDataSource = ExternalData;
  }
  __finally {
    PivotGrid()->EndUpdate();
    PivotGrid()->ApplyBestFit();
  }
}
//---------------------------------------------------------------------------

void __fastcall TfrmUnboundExternal::FormDestroy(TObject *Sender)
{
  delete ExternalData;
}
//---------------------------------------------------------------------------

TcxCustomPivotGrid* __fastcall TfrmUnboundExternal::PivotGrid()
{
  return UnboundPivotGrid;
}
//---------------------------------------------------------------------------

void __fastcall TfrmUnboundExternal::SetFieldPos(
  const String AFieldName, TcxPivotGridFieldArea AArea)
{
  TcxPivotGridField* AField = PivotGrid()->GetFieldByName(AFieldName);
  AField->Area = AArea;
}

//  TcxExternalDataSource

String TcxExternalDataSource::GetFieldName(int AIndex)
{
  return GetValueFromString(Records->Strings[0], AIndex);
}

String TcxExternalDataSource::GetFieldType(int AIndex)
{
  return GetValueFromString(Records->Strings[1], AIndex);
}

__fastcall TcxExternalDataSource::TcxExternalDataSource(const String FileName)
  : TcxCustomDataSource()
{
  Records = new TStringList();
  Records->LoadFromFile(FileName);
  AfterLoad();
}

__fastcall TcxExternalDataSource::~TcxExternalDataSource(void)
{
  delete Records;
}

void __fastcall TcxExternalDataSource::AfterLoad()
{
  String AItem;
  String S = Records->Strings[0];
  FFieldCount = 0;
  while (true) {
    FFieldCount = FFieldCount + 1;
    AItem = GetItemFromStr(S);
    S = S.SubString(AItem.Length() + 2, MaxInt);
    if ((AItem.Length() == 0) || (S.Length() == 0)) break;
  }
}

int __fastcall TcxExternalDataSource::GetRecordCount()
{
  return Records->Count - 2;
}

Variant __fastcall TcxExternalDataSource::GetValue(void * ARecordHandle, void * AItemHandle)
{
  Variant AResult = GetValueFromString(Records->Strings[(int)ARecordHandle + 2], (int)AItemHandle);
  return  AResult.AsType(GetVarTypeByName(FieldTypes[(int)AItemHandle]));
}

Variant __fastcall TcxExternalDataSource::GetValueFromString(String ARecord, int AIndex)
{
  String S = "";
  for (int I = 0; I <= AIndex; I++) {
    S = GetItemFromStr(ARecord);
    ARecord = ARecord.SubString(S.Length() + 2, MaxInt);
  }
  return S;
}

int __fastcall TcxExternalDataSource::GetVarTypeByName(const String AName)
{
  int AResult;
  if (SameText(AName, "Currency"))
    AResult = varCurrency;
  else
    if (SameText(AName, "DateTime"))
      AResult = varDate;
    else
      if (SameText(AName, "Integer"))
        AResult = varInteger;
      else
        AResult = varString;
  return AResult;
}

String __fastcall TcxExternalDataSource::GetItemFromStr(const String AString)
{
  int ACount = 0;
  for (int I = 1; I <= AString.Length(); I++) {
    if ((AString[I] == '\t') || (AString[I] == '\n') || (AString[I] == '\r'))
      break;
    else
      ACount = ACount + 1;
  }
  return AString.SubString(1, ACount);
}

//---------------------------------------------------------------------------

File: Components/DevExpressVCL/Demos/ExpressPivotGrid/Delphi/UnboundExternalDemo/UnboundExternalMain.pas
unit UnboundExternalMain;

{$I cxVer.inc}

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, DemoBasicMain, cxLookAndFeels, Menus, StdCtrls, cxControls,
  cxCustomPivotGrid, cxPivotGrid, cxCustomData, cxClasses, cxGraphics,
  cxStyles, cxLookAndFeelPainters, cxEdit;

type
  TcxExternalDataSource = class;
  TfrmUnboundExternal = class(TfrmDemoBasicMain)
    UnboundPivotGrid: TcxPivotGrid;
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
  protected
    ExternalData: TcxExternalDataSource;
    function GetPivotGrid: TcxCustomPivotGrid; override;
    procedure SetFieldPos(const AFieldName: string; AArea: TcxPivotGridFieldArea);
  public
    { Public declarations }
  end;

  TcxExternalDataSource = class(TcxCustomDataSource)
  private
    FFieldCount: Integer;
    function GetFieldName(AIndex: Integer): string;
    function GetFieldType(AIndex: Integer): string;
    function GetItemFromStr(const AString: string): string;
  protected
    Records: TStringList;
    procedure AfterLoad;
    function GetRecordCount: Integer; override;
    function GetValue(ARecordHandle: TcxDataRecordHandle; AItemHandle: TcxDataItemHandle): Variant; override;
    function GetValueFromString(ARecord: string; AIndex: Integer): Variant;
    function GetVarTypeByName(const AName: string): Integer; 
  public
    constructor Create(const AFileName: string);
    destructor Destroy; override;

    property FieldCount: Integer read FFieldCount;
    property FieldNames[AIndex: Integer]: string read GetFieldName;
    property FieldTypes[AIndex: Integer]: string read GetFieldType;
    property RecordCount: Integer read GetRecordCount;
  end;

var
  frmUnboundExternal: TfrmUnboundExternal;

implementation

{$R *.dfm}

uses
  dxCore;

function TfrmUnboundExternal.GetPivotGrid: TcxCustomPivotGrid;
begin
  Result := UnboundPivotGrid;
end;

procedure TfrmUnboundExternal.SetFieldPos(
  const AFieldName: string; AArea: TcxPivotGridFieldArea);
var
  AField: TcxPivotGridField; 
begin
  AField := PivotGrid.GetFieldByName(AFieldName);
  AField.Area := AArea;
end;

procedure TfrmUnboundExternal.FormCreate(Sender: TObject);
var
  I: Integer;
begin
  inherited;
  ExternalData := TcxExternalDataSource.Create('ExternalData.dat');
  PivotGrid.BeginUpdate;
  try
    for I := 0 to ExternalData.FieldCount - 1 do
    begin
      with PivotGrid.CreateField do
      begin
        Caption := ExternalData.FieldNames[I];
        DataBinding.ValueType := ExternalData.FieldTypes[I];
        Visible := True;
      end;
    end;
    SetFieldPos('PaymentType', faColumn);
    SetFieldPos('Payment Amount', faData);
    SetFieldPos('Quantity', faData);
    SetFieldPos('Company Name', faRow);
    SetFieldPos('Car Name', faRow);
    PivotGrid.DataController.CustomDataSource := ExternalData;
  finally
    PivotGrid.EndUpdate;
    PivotGrid.ApplyBestFit;
  end;
end;

procedure TfrmUnboundExternal.FormDestroy(Sender: TObject);
begin
  ExternalData.Free; 
end;

{  TcxExternalDataSource }

constructor TcxExternalDataSource.Create(const AFileName: string);
begin
  Records := TStringList.Create;
  Records.LoadFromFile(AFileName);
  AfterLoad;
end;

destructor TcxExternalDataSource.Destroy;
begin
  Records.Free;
  inherited Destroy;
end;

procedure TcxExternalDataSource.AfterLoad;
var
  S, AItem: string;
begin
  S := Records[0];
  FFieldCount := 0;
  repeat
    Inc(FFieldCount);
    AItem := GetItemFromStr(S);
    S := Copy(S, Length(AItem) + 2, MaxInt);
  until (Length(AItem) = 0) or (Length(S) = 0);
end;

function TcxExternalDataSource.GetRecordCount: Integer;
begin
  Result := Records.Count - 2;
end;

function TcxExternalDataSource.GetValue(
  ARecordHandle: TcxDataRecordHandle; AItemHandle: TcxDataItemHandle): Variant;
begin
  Result := GetValueFromString(Records[Integer(ARecordHandle) + 2], Integer(AItemHandle));
  VarCast(Result, Result, GetVarTypeByName(FieldTypes[Integer(AItemHandle)]));
end;

function TcxExternalDataSource.GetValueFromString(
  ARecord: string; AIndex: Integer): Variant;
var
  S: string;
  I: Integer;
begin
  S := '';
  for I := 0 to AIndex do
  begin
    S := GetItemFromStr(ARecord);
    ARecord := Copy(ARecord, Length(S) + 2, MaxInt);
  end;
  Result := S;
end;

function TcxExternalDataSource.GetVarTypeByName(const AName: string): Integer;
begin
  if SameText(AName, 'Currency') then
    Result := varCurrency
  else
    if SameText(AName, 'DateTime') then
      Result := varDate
    else
      if SameText(AName, 'Integer') then
        Result := varInteger
      else
        Result := varString;
end;

function TcxExternalDataSource.GetFieldName(AIndex: Integer): string;
begin
  Result := GetValueFromString(Records[0], AIndex);
end;

function TcxExternalDataSource.GetFieldType(AIndex: Integer): string;
begin
  Result := GetValueFromString(Records[1], AIndex);
end;

function TcxExternalDataSource.GetItemFromStr(const AString: string): string;
var
  I, ACount: Integer;
begin
  ACount := 0;
  for I := 1 to Length(AString) do
  begin
    if dxCharInSet(AString[I], [#9, #10, #13]) then
      Break
    else
      Inc(ACount);
  end;
  Result := Copy(AString, 1, ACount);
end;

end.
